File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

num, lim = map(int, sys.stdin.readline().split())
things = []
dp = [[0 for lim in range(lim+1)] for num in range(num)]
for _ in range(num):
    w, v = map(int, sys.stdin.readline().split())
    things.append([w, v])

# print(dp)

#DP[i][j]=max((DP[i][j-해당차례 물건 무게]+해당차례 물건 가치),DP[i-1][j])
for i in range(0, num):
    weight, value = things[i][0], things[i][1]
    for j in range(0, lim+1):
        if(j < things[i][0]): dp[i][j] = dp[i-1][j]
        else: dp[i][j] = max(dp[i-1][j], dp[i-1][j - weight] + value)
print(dp[num-1][lim])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String[] ipts;

        ipts = br.readLine().split(" ");
        int N = Integer.parseInt(ipts[0]);
        int K = Integer.parseInt(ipts[1]);

        int[] productWeights = new int[N];
        int[] productValues = new int[N];

        for (int i = 0; i < N; i++) {
            ipts = br.readLine().split(" ");
            productWeights[i] = Integer.parseInt(ipts[0]);
            productValues[i] = Integer.parseInt(ipts[1]);
        }

        // N개를 선택하며 K 무게만큼 담는 최대 가치
        int[][] dp = new int[N+1][K+1];
        int answer = 0;

        for (int i = 1; i <= N; i++) {
            int weight = productWeights[i-1];
            int value = productValues[i-1];

            for (int j = 1; j <= K; j++) {
                dp[i][j] = Math.max(dp[i][j], dp[i-1][j]);

                if(j >= weight){
                    dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j - weight] + value);
                    answer = Math.max(answer, dp[i][j]);                }
            }
        }

        System.out.println(answer);
    }
}
